### ISSUE TITLE
No method in multimethod 'connection-details->spec' for dispatch value: :clickhouse

### ISSUE THREAD

**Issue by MLikeWater** [enqueue/metabase-clickhouse-driver#43]
## Clickhouse and metabase version
Clickhouse: 19.17.2.4
Metabase: v0.34.0 (Latest)
metabase-clickhouse-driver: 0.6

## Deployment
I copy the driver jar clickhouse.metabase-driver.jar to the plugins directory, as follows:
```
# ll plugins/
total 117740
-rw-r--r-- 1 root root   392466 Dec 23 17:39 bigquery.metabase-driver.jar
-rw-r--r-- 1 root root  3875030 Dec 23 18:26 clickhouse.metabase-driver.jar
-rw-r--r-- 1 root root   465476 Dec 23 17:39 druid.metabase-driver.jar
-rw-r--r-- 1 root root   617794 Dec 23 17:39 googleanalytics.metabase-driver.jar
-rw-r--r-- 1 root root  3055790 Dec 23 17:39 google.metabase-driver.jar
-rw-r--r-- 1 root root  6550563 Dec 23 17:39 mongo.metabase-driver.jar
-rw-r--r-- 1 root root    68018 Dec 23 17:39 oracle.metabase-driver.jar
-rw-r--r-- 1 root root   140859 Dec 23 17:39 presto.metabase-driver.jar
-rw-r--r-- 1 root root  2381192 Dec 23 17:39 redshift.metabase-driver.jar
-rw-r--r-- 1 root root 30587032 Dec 23 17:39 snowflake.metabase-driver.jar
-rw-r--r-- 1 root root 63997026 Dec 23 17:39 sparksql.metabase-driver.jar
-rw-r--r-- 1 root root  7114948 Dec 23 17:39 sqlite.metabase-driver.jar
-rw-r--r-- 1 root root  1229843 Dec 23 17:39 sqlserver.metabase-driver.jar
-rw-r--r-- 1 root root    50615 Dec 23 17:39 vertica.metabase-driver.jar
```

Start Metabase:
```
MB_PLUGINS_DIR=./plugins; nohup  java -jar metabase.jar &
```

## Error Info
```
12-23 18:27:55 ERROR driver.util :: Database connection error
java.lang.IllegalArgumentException: No method in multimethod 'connection-details->spec' for dispatch value: :clickhouse
        at clojure.lang.MultiFn.getFn(MultiFn.java:156)
        at clojure.lang.MultiFn.invoke(MultiFn.java:233)
        at metabase.driver.sql_jdbc.connection$details__GT_connection_spec_for_testing_connection.invokeStatic(connection.clj:152)
        at metabase.driver.sql_jdbc.connection$details__GT_connection_spec_for_testing_connection.invoke(connection.clj:147)
        at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:158)
        at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:154)
        at metabase.driver.sql_jdbc$fn__69874.invokeStatic(sql_jdbc.clj:35)
        at metabase.driver.sql_jdbc$fn__69874.invoke(sql_jdbc.clj:34)
        at clojure.lang.MultiFn.invoke(MultiFn.java:234)
        at metabase.driver.util$can_connect_with_details_QMARK_$fn__21110.invoke(util.clj:32)
        at metabase.util$do_with_timeout$fn__13131.invoke(util.clj:302)
        at clojure.core$binding_conveyor_fn$fn__5754.invoke(core.clj:2030)
        at clojure.lang.AFn.call(AFn.java:18)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
12-23 18:27:55 DEBUG middleware.log :: POST /api/database 400 28.3 ms (0 DB calls) 
{:valid false,
 :dbname "No method in multimethod 'connection-details->spec' for dispatch value: :clickhouse",
 :message "No method in multimethod 'connection-details->spec' for dispatch value: :clickhouse"}
```

**Comment by MLikeWater** [comment]
I have downloaded metabase enterprise version, and I put the jar package in the plugins directory.
As a result It works well.

**Comment by yarhrn** [comment]
I have a similar problem running a docker container build from the docker file in the readme.
Also, there is an error during driver initialization 
```
Caused by: java.io.FileNotFoundException: Could not locate metabase/util/date__init.class, metabase/util/date.clj or metabase/util/date.cljc on classpath.
```

**Comment by yarhrn** [comment]
UPD:
I managed to add clickhouse db by downgrading to metabase v0.33

**Comment by enqueue** [comment]
The Metabase 0.34 release has brought many changes to fix timezone related issues (see [Release Announcement](https://www.metabase.com/blog/0.34/index.html) and [issue list](https://github.com/metabase/metabase/issues?q=is%3Aissue+label%3AMisc%2FTimezones+is%3Aclosed+milestone%3A0.34.0). This is a good thing™, but the drivers need to be updated.

**Comment by pyrooka** [comment]
The date utils has completely rewritten and renamed from `metabase.util.date` to `metabase.util.date-2`. I faced this problem with my driver too. Fortunately, I only had to change the name and some functions.

**Comment by yarhrn** [comment]
Got it, thanks!!!

**Comment by argnist** [comment]
Is there a workaround except downgrading to 0.33?

**Comment by pyrooka** [comment]
> Is there a workaround except downgrading to 0.33?

I think, no.

**Comment by argnist** [comment]
Unfortunately we have the same error with Oracle in 0.33.x.

**Comment by enqueue** [comment]
@argnist the Oracle driver is supported by the Metabase team, and it would be strange if it did not work. Please check the [support forum](https://discourse.metabase.com).

Meanwhile, I spent some time on the ClickHouse driver. Unfortunately, toggling the "legacy mode" for drivers, as [hinted at](https://github.com/dacort/metabase-athena-driver/issues/27#issuecomment-571282399) by @camsaul is not a viable path for our driver, because the relevant JDBC methods, i.e. setting of date/time parameters with an optional `Calendar` instance to hold time zone, aren't implemented, either.

I am trying to build support for `java.time` into the JDBC driver, but it will take some time (see ClickHouse/clickhouse-jdbc#406)

**Comment by pyrooka** [comment]
I guess this is what caused the Oracle driver problem: https://github.com/metabase/metabase/issues/11699

**Comment by long2ice** [comment]
hello,any news?I really want use clickhouse in metabase,wonder why official support for clickhouse is not active...

**Comment by enqueue** [comment]
Sorry there isn't much news. I started working on the ClickHouse JDBC driver to include support for java.time data types. I do not know when the PR is ready for review, yet.

**Comment by enqueue** [comment]
If you are adventurous, feel free to try out 0.7 BETA release, and report your findings. We are particularly interested in reports regarding

* grouping date time fields by dates, using different time zones in JVM, ClickHouse server, client
* aggregating / filtering on Date columns

The driver package includes my current fork of the JDBC driver (see branch [jdbc_4.2](https://github.com/enqueue/clickhouse-jdbc/tree/jdbc_4.2)) and a couple of adjustments of the actual Metabase ClickHouse driver.

Using the official test suite, we are currently down to 1 failure and 3 errors, which have to be fixed.

**Comment by skaboy007** [comment]
I tested it by making a setup copy of my existing 0.6 + 0.33.6 to 0.7beta + 0.34.2 
seems to be working fine. filter issues seem to be gone. did not see any issues at the moment. will be running it as if production for a few days. my case is fault-tolerant + I have the 0.6 + 0.33.6 running also.

**Comment by dangerous1990** [comment]
> I tested it by making a setup copy of my existing 0.6 + 0.33.6 to 0.7beta + 0.34.2
> seems to be working fine. filter issues seem to be gone. did not see any issues at the moment. will be running it as if production for a few days. my case is fault-tolerant + I have the 0.6 + 0.33.6 running also.

just replace clickhouse jar 0.6 to 0.7beta?

**Comment by skaboy007** [comment]
> > I tested it by making a setup copy of my existing 0.6 + 0.33.6 to 0.7beta + 0.34.2
> > seems to be working fine. filter issues seem to be gone. did not see any issues at the moment. will be running it as if production for a few days. my case is fault-tolerant + I have the 0.6 + 0.33.6 running also.
> 
> Just replace clickhouse jar 0.6 to 0.7beta?

No, I made a full copy of the setup just in case.

**Comment by enqueue** [comment]
I released 0.7.0 including a fix for this issue.

I am not happy about the current state of the driver because it relies on a patched version of the official ClickHouse JDBC driver. But the driver should work fine with Metabase 0.34.x We have run 0.7-beta2 for a couple of weeks now without major problems.

**Comment by vini-intenseye** [comment]
Hi i have the same issue with the latest metabase.jar
Metabase version: 0.44.2

Any hints?

Error
2023-11-13 13:06:04,290 DEBUG middleware.log :: POST /api/database 400 6.6 ms (0 DB calls) 
{:message "No method in multimethod 'connection-details->spec' for dispatch value: :clickhouse"}

**Comment by slvrtrn** [comment]
@vini-intenseye, the latest driver jar won't work on 0.44.2; there are too many breaking changes across the Metabase versions. See https://github.com/ClickHouse/metabase-clickhouse-driver#choosing-the-right-version

**Comment by vini-intenseye** [comment]
I just downgrade the plugin to get it working. thanks

**Comment by wanggi404** [comment]
2023-12-05T19:24:57+07:00 DEBUG metabase.server.middleware.log POST /api/database 400 5.7 ms (1 DB calls) 
{:message "No method in multimethod 'connection-details->spec' for dispatch value: :clickhouse"}

i have same issue with metabase enterprise version v1.47.4,

do you guys have any recomend solution to fix this issue guys? im using driver v1.2.5

**Comment by wanggi404** [comment]
> 2023-12-05T19:24:57+07:00 DEBUG metabase.server.middleware.log POST /api/database 400 5.7 ms (1 DB calls) {:message "No method in multimethod 'connection-details->spec' for dispatch value: :clickhouse"}
> 
> i have same issue with metabase enterprise version v1.47.4,
> 
> do you guys have any recomend solution to fix this issue guys? im using driver v1.2.5

sorry i didn't read carefully, im using driver v1.23 and now work perfectly
